### ISSUE TITLE
Feature: custom proxy settings

### ISSUE THREAD

**Issue by yianL** [kadirahq/react-storybook#208]
First of all, great work with react-story book! I love how I can now jump between different states of my components quickly, and work on the visual changes.

So I have a component that fetches data from an api, and plots it onto a chart. Although I can write a story and feed mocked data into the component, sometimes it's helpful to render the "smart" version of the component and see it in real action. Problem is, I used to setup my own express server with webpackDevMiddleware, and then configure proxies so I can redirect api calls to another port. I couldn't find a way to setup similar configuration when using react-storybook. I ended up with a hack by going into `node_modules/.bin/start-storybook` and put my proxy settings there.

I'm not sure if this is a use case the authors intended to support, but I think it'll be really nice if I can use react-storybook as a playground for all sorts of components, whether they're dumb or smart. To achieve that, we would need a way to support custom proxy settings.

**Comment by mnmtanish** [comment]
Hi @yianL,

Thanks for bringing this use case to our attention. I started working on a way to make storybook more flexible [here](https://github.com/kadirahq/react-storybook/pull/211).

**Comment by mnmtanish** [comment]
Published `v1.28.0`. You can now import the middleware and write your own express server.

```js
import storybook from '@kadira/storybook/dist/server/middleware';

// ...
app.use(storybook(configDirectory));
// ...
```

Check this [line](https://github.com/kadirahq/react-storybook/blob/3a21d57d30f7a7f0252205f9dfc6dc820fb46e94/src/server/index.js#L49)

**Comment by yianL** [comment]
Wow, this is awesome. I will definitely try it out. Feel free to close this issue.

**Comment by maxww** [comment]
@yianL do you mind sharing the config sample you are using that works with storybook?

**Comment by ndelangen** [comment]
@maxww What kind of example are you looking for?
Something like this:
https://github.com/mthuret/storybook-addon-specifications/blob/jest-test-files/.storybook/middleware.js ?

**Comment by ndelangen** [comment]
We should document this ability better I think.

**Comment by maxww** [comment]
@ndelangen I don't have server side and one of the components is using xhr to make ajax call get fetch more data. I want to write the webpack.config to be able to do like 
 ```
 proxy: {
      '/pws': 'http://www.someline.com'
}
```

I've tried it to be inside and outside the devSever object, no luck.
Since I'm fairly new to this, I'm not sure if I was even doing this right.

**Comment by yianL** [comment]
@maxww Here's a sample of my setup:

```js
var Express = require('express')
var proxy = require('http-proxy-middleware')
var storybook = require('@kadira/storybook/dist/server/middleware').default
var app = new Express()
var port = 4001
var storybookConfig = './.storybook'

app.use(storybook(storybookConfig))

app.use('/api', proxy({
  target: 'https://api-endpoint', // target host
  changeOrigin: true,
}))

app.listen(port, function (error) {
  if (error) {
    console.error(error)
  } else {
    console.info('==> 🌎  Listening on port %s. Open up http://localhost:%s/ in your browser.', port, port)
  }
})
```

Let me know if you have any questions.

**Comment by ndelangen** [comment]
I have to say that's some creative usage, but I cannot recommend anyone to use files deep into an package. 

We cannot and will not guarantee those files will stay in those location or contain the same exports. It's not our API.

Including storybook inside an express app is a use case ew haven't seen much so far; though I'm interested in supporting it officially at some point.

@maxww storybook HAS a 'backend' and it's express. By adding a `middleware.js` in your storybook config folder, you can inject any route you want.

I guess you can try and do the same inside a webpack config, I don't know I've never tried that.
Let me know when you pull it off, if that works it'd be great to document it how.

**Comment by ndelangen** [comment]
I don't understand how you say you don't have a backend, but you component is fetching data somewhere?

Do you mean you're building storybook to a static site?

**Comment by indiesquidge** [comment]
I'm not sure if this is what @maxww was getting at, but I'm wondering about a similar situation in which I "don't have a backend", meaning that I am calling an external API for my data.


## TL;DR

So my question would be this: _is there a way for me to initialize Storybook with an API proxy without having a local Express server or `webpack.config.js` to customize myself?_


## Initializing Storybook with an API Proxy

I am using [Create React App](https://github.com/facebookincubator/create-react-app/) for my React project now, and would love to find a way in which I can still call the API endpoints when using Storybook.

CRA makes use of a `proxy` property in the `package.json` to allow for [proxying API requests in development](https://github.com/facebookincubator/create-react-app/blob/master/packages/react-scripts/template/README.md#proxying-api-requests-in-development). Something similar to what @maxww posted.

> ⚠️ Note: CRA does not expose a webpack config file directly, as one of it's [core philosophies](https://github.com/facebookincubator/create-react-app#philosophy) is "no configuration required". Because of this, writing custom webpack config is not an option unless one [ejects](https://github.com/facebookincubator/create-react-app/blob/master/packages/react-scripts/template/README.md#npm-run-eject) from CRA, which may be unavoidable once a team wants a more granular level of control and customization, but is not recommended unless absolutely needed since it is irreversible.

Ideally, I would love to keep this working with isolated components in Storybook, so that things like an autocomplete search form can be properly used, styled, and tested.

I currently have this working by using [Storybook environment variables](https://storybook.js.org/configurations/env-vars/). (As an aside, I don't know if the documentation for environment variables is still actively being kept up-to-date, as I only found it via a Google search, and it does not appear anywhere as a clickable link within the Storybook column nav panel.)

```javascript
const apiUrl = process.env.REACT_APP_API_URL || process.env.STORYBOOK_API_URL

export const searchSuggestions = fragment =>
  Get(`${apiUrl}/autocomplete/v1/suggestions/${fragment}`)

// ...
```

I then start Storybook like

```shell
STORYBOOK_API_URL="https://example.com" yarn storybook
```

And now I can make calls the the API endpoints.

However, this feels a bit hacky, and goes against what the Storybook environment variables docs recommend

> _Even though we can access these env variables anywhere in the client side JS code, it’s better to use them only inside stories and inside the main Storybook config file._

Is it possible to initialize Storybook with an API proxy without having to run my own server or edit the webpack config directly? Perhaps some form of middleware in the Storybook setup?

If this is not possible, then are there any foreseeable repercussions with the current method I have set up that uses a Storybook environment variable?

**Comment by ndelangen** [comment]
@indiesquidge wanna do a 1 on 1 talk about thi topic in a sky call or something? talk to me on slack 👍

**Comment by indiesquidge** [comment]
@ndelangen, just when I saw your message I was able to find a solution to my problem using the awesome middleware customization roughly outlined in #435.

I have the following in `.storybook/middleware.js`

```javascript
const proxy = require('http-proxy-middleware')

module.exports = function expressMiddleWare (router) {
  router.use('/api', proxy({
    target: 'https://api-endpoint.com',
    changeOrigin: true
  }))
}
```

This has a dependency on the `http-proxy-middleware` package.

It uses a proxy function to map a context ("/api") to a target ("https://api-endpoint.com"). One thing to note is that the `changeOrigin` option must be explicitly set to true in order for the origin of the host header to be changed to the target URL. This can be done for any number of proxies.

Thank y'all so much for exposing the middleware custom config in Storybook, that is a very helpful feature indeed 💛

**Comment by shilman** [comment]
@indiesquidge can you explain your workflow a bit more? so you have stories for container-style components that are end-to-end connected to your live API? Or do your stories grab the data from the server and then use it populate your components? Do you have ay public examples of stories like this I can look at, either in a repo or just comment here or create a gist? Super curious.

**Comment by indiesquidge** [comment]
> you have stories for container-style components that are end-to-end connected to your live API

Yeah, this is more inline with what I am working with.

Unfortunately I don't have a public example to show, but the main component that was causing me grief was an `<AutocompleteForm />` component that makes a request to a backend API for autocomplete suggestions based on the user's query fragment on every keypress. The suggestions are stored in the component state, and I am using the API to update the state inside of a component method.

Something like this

```javascript
export default class AutocompleteForm extends Component {
  state = {
    fragment: '',
    suggestions: []
  }

  handleSearch = fragment => {
    return api
      .searchSuggestions(fragment)
      .then(api.validateResponse)
      .then(api.toJSON)
      .then(json => json.suggestions.map(s => s.term))
      .then(suggestions => this.setState(() => ({ suggestions })))
      .catch(() => this.setState(() => ({ suggestions: [] })))
  }

  onChange = e => {
    const fragment = e.target.value
    this.handleSearch(fragment)
    this.setState(() => ({ fragment }))
  }

  render () {
    const { fragment, suggestions } = this.state

    return (
      <form>
        <label htmlFor='product-search'>Search</label>
        <input id='product-search' value={fragment} onChange={this.onChange} />
      </form>
      <ul>
        {suggestions && suggestions.map((suggestion, i) => <li key={i}>{suggestion}</li>}
      </ul>
    )
  }
}
```

The real component is a bit more complex, but this is the gist of it. I'd like to be able to show the suggestions list of `<AutocompleteForm />` in the story.

`api.searchSuggestions` looks similar to how I wrote it in [my first comment](https://github.com/storybooks/storybook/issues/208#issuecomment-306901362), only I'm not using Storybook environment variables anymore.

```javascript
const apiUrl = process.env.REACT_APP_API_URL || ''

export const searchSuggestions = fragment =>
  fetch(`${apiUrl}/api/autocomplete/v1/suggestions/${fragment}`)
```

**Comment by shilman** [comment]
@indiesquidge Thanks for sharing! I think this would make an awesome blog post once you're happy with a solution. Would be thrilled to publish it on Storybook's Medium publication if that's at all interesting: https://medium.com/storybookjs

**Comment by indiesquidge** [comment]
@shilman, that sounds great! I'll work on something this week and get back to you soon :)

**Comment by djyde** [comment]
It's really a useful feature, since the `webpack-dev-server` has a `proxy` config out of the box. Though we should do this work manually, but why there is no document for it? I don't even know the middle ware can be override by adding a `middleware.js`.

I am willing to create a PR for this part of doc.

**Comment by ndelangen** [comment]
A PR on our docs to improve the information regarding `middleware.js` would be fantastic!

**Comment by djyde** [comment]
I will work on it. 😋

**Comment by hmontes** [comment]
Any news? I want to connect Apollo Graphql and i need this option (Like proxy config in create-react-app).

Thank you :)

**Comment by djyde** [comment]
@hmontes This is how I did:

```js
// .storybook/middleware.js

const proxy = require('http-proxy-middleware')
const packageJson = require('../package.json')

module.exports = function expressMiddleware(router) {
  const proxyConfig = packageJson.proxy || {}

  for (let domain in proxyConfig) {
    if (typeof proxyConfig[domain] === 'string') {
      console.log(domain)
      router.use(domain, proxy({
        target: proxyConfig[domain]
      }))
    } else {
      router.use(domain, proxy(proxyConfig[domain]))
    }
  }
}
```

```json
// package.json

{
  "proxy": {
    "/api": "http://localhost:5000/api"
  }
}
```

**Comment by adamchenwei** [comment]
anyone would add the doc for allowing webpack dev server proxy into the storybook?

**Comment by adamchenwei** [comment]
this should add into the doc...

```
const proxy = require('http-proxy-middleware')

module.exports = function expressMiddleware (router) {
  router.use('/api', proxy({
    target: 'https://api-endpoint.com',
    changeOrigin: true
  }))
}
```

**Comment by chrbala** [comment]
To make it API-compatible with CRA, the following works:

```javascript
const proxy = require('http-proxy-middleware');

const { proxy: proxyConfigs } = require('../package.json');

module.exports = router =>
  Object.keys(proxyConfigs).forEach(key =>
    router.use(key, proxy(proxyConfigs[key]))
  );
```

IMO this should be added as a default to storybook, since it does say that it bases the default setup on CRA.

**Comment by adamchenwei** [comment]
@chrbala  what is CRA?

**Comment by ndelangen** [comment]
[Create React App](https://github.com/facebookincubator/create-react-app)

**Comment by jakiestfu** [comment]
Thanks for this, super flexible! Any way to get something like this to work with `build-storybook`?

**Comment by japboy** [comment]
It is cool if `devServer.proxy` would work just like `webpack-dev-server`. I'm using Storybook w/ Vue

**Comment by angeliski** [comment]
Hey @japboy I create a wordaround for that case, you can see here:
https://github.com/angeliski/storybook-vue-webpack-proxy
Basicaly I use the webpack to route to storybook. It's ugly, but for now works.

@ndelangen You have some improve in mind for that situation?

**Comment by picheli20** [comment]
Who's using Angular, can use what @djyde did but simply use the `.storybook/middleware.js` and use your current `proxy.config.json`:

```
const proxy = require('http-proxy-middleware')
const proxyConfig = require('../proxy.conf.json')

module.exports = function expressMiddleware(router) {

  Object.keys(proxyConfig).forEach(domain => {
    const target = proxyConfig[domain];
    if (typeof target === 'string') {
      console.log(domain)
      router.use(domain, proxy({ target }))
    } else {
      router.use(domain, proxy(proxyConfig[domain]))
    }
  });
}
```

**Comment by dereklin** [comment]
I am using Angular.  Is there a way to do proxy per story instead of per the whole storybook?

**Comment by picheli20** [comment]
@dereklin you can add some prefix for earch story on the API request and re-write it on the proxy it-self :)

**Comment by dtslvr** [comment]
> Who's using Angular, can use what @djyde did but simply use the `.storybook/middleware.js` and use your current `proxy.config.json`:
> 
> ```ts
> const proxy = require('http-proxy-middleware')
> const proxyConfig = require('../proxy.conf.json')
> 
> module.exports = function expressMiddleware(router) {
> 
>   Object.keys(proxyConfig).forEach(domain => {
>     const target = proxyConfig[domain];
>     if (typeof target === 'string') {
>       console.log(domain)
>       router.use(domain, proxy({ target }))
>     } else {
>       router.use(domain, proxy(proxyConfig[domain]))
>     }
>   });
> }
> ```

Hi @picheli20

Thanks for the snippet. I have copied it to `.storybook/middleware.js`. In the terminal I can see `[HPM] GET /api/v1/users/59b49fbc-022e-47af-8469-b88f9158a36b:2ac82679-f51a-4524-8684-13132d2a180d -> http://localhost:8888`, but in _DevTools_ > _Network_  there is still `GET http://localhost:6006/api/v1/users/59b49fbc-022e-47af-8469-b88f9158a36b:2ac82679-f51a-4524-8684-13132d2a180d` (`401 Unauthorized`) .

Any ideas?

**Comment by picheli20** [comment]
@dtslvr and the same request to the backend works? Looks like that the request is going wrong, not the proxy

**Comment by machineghost** [comment]
A minor update: all of the solutions here rely on the default export of the proxy middleware package ... but that didn't work for me.

The package now seems to be relying on a `createProxyMiddleware` *named* export instead, so what worked for me was:

```
// Add proxy of /images to localhost:3000/images
const { createProxyMiddleware } = require('http-proxy-middleware');

module.exports = router => {
  const proxyMiddleware = createProxyMiddleware('/images', {
    target: 'https://localhost:3000/images',
  });
  router.use(proxyMiddleware);
  return router;
};
```

**Comment by masood1** [comment]
> @ndelangen, just when I saw your message I was able to find a solution to my problem using the awesome middleware customization roughly outlined in #435.
> 
> It is a very similar solution to @yianL's solution [above](https://github.com/storybooks/storybook/issues/208#issuecomment-305567726), only I am not creating my own Express server and instead leveraging Storybook's router handler in the middleware.
> 
> I have the following in `.storybook/middleware.js`
> 
> ```js
> const proxy = require('http-proxy-middleware')
> 
> module.exports = function expressMiddleware (router) {
>   router.use('/api', proxy({
>     target: 'https://api-endpoint.com',
>     changeOrigin: true
>   }))
> }
> ```
> 
> This has a dependency on the [`http-proxy-middleware`](https://github.com/chimurai/http-proxy-middleware) package.
> 
> It uses a proxy function to map a context ("/api") to a target ("api-endpoint.com"). One thing to note is that the `changeOrigin` option must be explicitly set to true in order for the origin of the host header to be changed to the target URL. The proxy function can be used for any number of proxies.
> 
> Thank y'all so much for exposing the middleware custom config in Storybook, that is a very helpful feature indeed 💛

Please use createProxyMiddleware, function is updated 

const { createProxyMiddleware } = require('http-proxy-middleware');
module.exports = function(app) {
  app.use(
    '/api/login',
    createProxyMiddleware({
      target: 'http://localhost:3090',
      changeOrigin: true,
    })
  );
};

**Comment by shilman** [comment]
FYI all of these APIs are experimental/undocumented, and we reserve the right to change/remove them outside of the normal rules of semver.

**Comment by teebodscvr** [comment]
> Who's using Angular, can use what @djyde did but simply use the `.storybook/middleware.js` and use your current `proxy.config.json`:
> 
> ```ts
> const proxy = require('http-proxy-middleware')
> const proxyConfig = require('../proxy.conf.json')
> 
> module.exports = function expressMiddleware(router) {
> 
>   Object.keys(proxyConfig).forEach(domain => {
>     const target = proxyConfig[domain];
>     if (typeof target === 'string') {
>       console.log(domain)
>       router.use(domain, proxy({ target }))
>     } else {
>       router.use(domain, proxy(proxyConfig[domain]))
>     }
>   });
> }
> ```

This worked for me in an Angular project

**Comment by federico-ntr** [comment]
> Please use createProxyMiddleware, function is updated
> 
> const { createProxyMiddleware } = require('http-proxy-middleware'); module.exports = function(app) { app.use( '/api/login', createProxyMiddleware({ target: 'http://localhost:3090', changeOrigin: true, }) ); };

This did the trick for me, but it doesn't work if it's a TypeScript file. I had to rename it from `middleware.ts` to `middleware.js`. Am I missing something or should I open a new issue?

Note: I'm using Storybook 7.0.0-rc.5 with vite

**Comment by Evgeny-Chugaev** [comment]
Works for me, "http-proxy-middleware": "^2.0.6",

`const proxy = require('http-proxy-middleware')

module.exports = function expressMiddleware(router) {
  router.use('/api', proxy.createProxyMiddleware({
    target: 'https://avtoliga.dev.design.ru',
    changeOrigin: true
  }))
}`

**Comment by ndelangen** [comment]
> I had to rename it from middleware.ts to middleware.js

Not a bug, it should be `middleware.js` exactly.
Improving the API isn't likely to happen, because of it's limitation that it can only work for development and not for build versions of storybook.

Long term we'd like to come up with a way to add serverside routes support to storybook that will be statically build.
But this API doesn't exist yet.

**Comment by biaogebusy** [comment]
working for me:
`const { createProxyMiddleware } = require("http-proxy-middleware");

module.exports = function expressMiddleware(router) {
	router.use(
		"/sites",
		createProxyMiddleware({
			target: "domain.com",
			changeOrigin: true,
		})
	);
};
`
